plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java

```diff
@@ -0,0 +1,195 @@
+    CloudSdkServer server = CloudSdkServer.getCloudSdkServer(getServer());
+    int port = server.getApiPort();
+    if (!isPortAvailable(port)) {
+      String message = server.getServer().getName() + " has admin port set to "
+                       + port
+                       + " which is already is in use.\nTo use port "
+                       + port
+                       + " stop the processes that are using it";
+      Activator.logError(message);
+      return new Status(IStatus.ERROR, Activator.PLUGIN_ID, message);
+    }
+    return Status.OK_STATUS;
+  }
+
+  @Override
+  public void stop(boolean force) {
+    terminate();
+  }
+
+  /**
+   * Notifies that the server started.
+   */
+  public void setServerStarted() {
+    setServerState(IServer.STATE_STARTED);
+  }
+
+  /**
+   * Finalizes preparations to launch server.
+   *
+   * @param launchMode the mode in which a server is running
+   */
+  protected void setupLaunch(String launchMode) {
+    setServerRestartState(false);
+    setServerState(IServer.STATE_STARTING);
+    setMode(launchMode);
+    String adminHost = CloudSdkServer.getCloudSdkServer(getServer()).getApiHost();
+    int adminPort = CloudSdkServer.getCloudSdkServer(getServer()).getApiPort();
+
+    // ping server to check for startup
+    String url = "http://" + adminHost + ":" + adminPort;
+    cloudSdkServerStartupChecker = new CloudSdkServerStartupChecker(getServer(), url, Integer.MAX_VALUE, this);
+    cloudSdkServerStartupChecker.start();
+  }
+
+  /**
+   * Terminates the Cloud SDK server instance, stops the ping thread and removes
+   * debug listener.
+   */
+  protected void terminate() {
+    int serverState = getServer().getServerState();
+    if ((serverState == IServer.STATE_STOPPED) || (serverState == IServer.STATE_STOPPING)) {
+      return;
+    }
+
+    try {
+      setServerState(IServer.STATE_STOPPING);
+      stopDevAppServer(); // sends a "quit" message to port
+      ILaunch launch = getServer().getLaunch();
+      if (launch != null) {
+        launch.terminate();
+      }
+
+      if (cloudSdkServerStartupChecker != null) {
+        cloudSdkServerStartupChecker.stop();
+        cloudSdkServerStartupChecker = null;
+      }
+
+      if (processListener != null) {
+        DebugPlugin.getDefault().removeDebugEventListener(processListener);
+        processListener = null;
+      }
+
+      // Terminate the remote debugger
+      CloudSdkServer server = CloudSdkServer.getCloudSdkServer(getServer());
+      ILaunchConfigurationWorkingCopy remoteDebugLaunchConfig = server.getRemoteDebugLaunchConfig();
+      if (remoteDebugLaunchConfig != null) {
+        ILaunchManager manager = DebugPlugin.getDefault().getLaunchManager();
+        for (ILaunch aLaunch : manager.getLaunches()) {
+          if (aLaunch.getLaunchConfiguration().equals(remoteDebugLaunchConfig)) {
+            try {
+              aLaunch.terminate();
+              manager.removeLaunch(aLaunch);
+              break;
```

Comment:
Looks like it's iterating over the launches, so there's a possibility of a `ConcurrentModificationException`.